File: chapter 2/js/2.3.js
// Delete Middle Node: Implement an algorithm to delete a node in the middle (i.e., any node but
//     the first and last node, not necessarily the exact middle) of a singly linked list, given only access to
//     that node. 
class Node{
    constructor(value){
        this.value = value;
        this.next = null 
    }
}

class SLL{
    constructor(){
        this.head = null;
    }

    append(val){
        let newNode = new Node(val);
        if (this.head == null){
            this.head = newNode
            return
        }
        let runner = this.head;
        while(runner.next != null){
            runner = runner.next;
        }   
        runner.next = newNode;
        return
    }
    printAll(){
        let runner = this.head
        while(runner != null){
            console.log(runner.value)
            runner = runner.next
        }
    }
    kthToLast(k){
        if(this.head == null || k == 0){
            return null
        }
        let first = this.head;
        let second = this.head;
        for(let i=1; i<=k; i++){
            if(first == null){
                return false
            }
            first = first.next;
        }
        while(first != null && second.next != null){
            first = first.next;
            second = second.next;
        }
        return second;
    }

    deleteMiddleNode(Node){
        if(this.head == null || this.head.next == null || this.head.next.next == null){
            return
        }
        let runner = this.head;
        while(runner.next.next != null){
            if(runner.next == Node){
                runner.next = runner.next.next
                return
            }
            runner = runner.next
        }
    }
    
}
let node = new Node(3)
console.log(node)
let sll = new SLL()
sll.append(2)
sll.append(3)
sll.append(4)
sll.append(5)
sll.append(6)
sll.append(7)
sll.append(3)
sll.append(5)
let middleNode = sll.kthToLast(8)
sll.deleteMiddleNode(middleNode)
sll.printAll()

File: chapter 2/python/2.3.py
# 2.3 Delete Middle Node: Implement an algorithm to delete a node in the middle (i.e., any node but
# the first and last node, not necessarily the exact middle) of a singly linked list, given only access to
# that node. 

class Node:
    def __init__(self, val, next=None):
        self.val = val
        self.next = next
    
class SLL:
    def __init__ (self):
        self.head = None

    def printAll(self):
        current = self.head
        while current != None:
            print(current.val)
            current = current.next
    
    def addNode(self, val):
        if self.head == None:
            self.head = Node(val)
        else:
            newNode = Node(val)
            current = self.head
            while current.next != None:
                current = current.next
            current.next = newNode
    
    def removeFromMiddle(self, val):
        if self.head == None or self.head.next == None:
            return False
        current = self.head
        while current.next != None:
            if current.next.val == val and current.next.next != None:
                current.next = current.next.next
                isRemoved = True
                return 
            current = current.next
        print("the value does not exist")
                

list = SLL()
list.addNode("a")
list.addNode("b")
list.addNode("c")
list.addNode("c")
list.addNode("d")
list.addNode("c")
list.addNode("e")
# list.printAll()
list.removeFromMiddle("c")
list.printAll()